File: python/src/gibberish_rewriter/win/layout_pair_resolver.py
"""The ``layouts`` setting: "auto" or explicit HKL values, checked against the installed layouts. Pure.

Mirrors ``csharp/src/GibberishRewriter.App/Win/LayoutPairResolver.cs``.
"""

from __future__ import annotations

import unicodedata
from collections.abc import Callable, Sequence
from dataclasses import dataclass

from gibberish_rewriter.app import app_messages
from gibberish_rewriter.win import hkl
from gibberish_rewriter.win.hkl import LayoutPair


@dataclass(frozen=True)
class PairResolution:
    """Either a pair or the notification that explains why there is none."""

    pair: LayoutPair | None
    error: str | None


def resolve(
    latin_setting: str, arabic_setting: str, installed: Sequence[tuple[int, str]]
) -> PairResolution:
    """``installed`` is each installed layout with what its KeyA types plain."""
    latin, latin_error = _pick(latin_setting, installed, _is_ascii_letter)
    if latin_error is not None:
        return PairResolution(None, latin_error)
    arabic, arabic_error = _pick(arabic_setting, installed, _is_arabic_letter)
    if arabic_error is not None:
        return PairResolution(None, arabic_error)
    if latin == arabic:
        return PairResolution(None, app_messages.LAYOUTS_SAME)
    return PairResolution(LayoutPair(latin, arabic), None)


def _pick(
    setting: str, installed: Sequence[tuple[int, str]], is_script_letter: Callable[[str], bool]
) -> tuple[int, str | None]:
    if setting == "auto":
        matches = [layout for layout, key_a in installed if len(key_a) == 1 and is_script_letter(key_a)]
        return (matches[0], None) if len(matches) == 1 else (0, app_messages.LAYOUTS_NOT_FOUND)
    # An eight-hex-digit setting that is not hex raises, exactly as C#'s uint.Parse does; ConfigParser has
    # already answered layout.invalid for that case.
    value = hkl.parse(setting)
    if any(layout == value for layout, _ in installed):
        return value, None
    return 0, app_messages.layout_not_installed(setting)


def _is_ascii_letter(c: str) -> bool:
    """C#'s ``char.IsAsciiLetter``."""
    return "A" <= c <= "Z" or "a" <= c <= "z"


def _is_arabic_letter(c: str) -> bool:
    """The same test as TextConverter: U+0600-U+06FF with a letter category."""
    return "\u0600" <= c <= "\u06ff" and unicodedata.category(c).startswith("L")

File: csharp/src/GibberishRewriter.App/Win/LayoutPairResolver.cs
using System.Globalization;

namespace GibberishRewriter.App.Win;

/// <summary>Either a pair or the notification that explains why there is none.</summary>
internal sealed record PairResolution(LayoutPair? Pair, string? Error);

/// <summary>The <c>layouts</c> setting: "auto" or explicit HKL values, checked against the installed layouts. Pure.</summary>
internal static class LayoutPairResolver
{
    /// <param name="installed">Each installed layout with what its KeyA types plain.</param>
    public static PairResolution Resolve(string latinSetting, string arabicSetting, IReadOnlyList<(uint Hkl, string KeyA)> installed)
    {
        var (latin, latinError) = Pick(latinSetting, installed, char.IsAsciiLetter);
        if (latinError is not null)
        {
            return new PairResolution(null, latinError);
        }
        var (arabic, arabicError) = Pick(arabicSetting, installed, IsArabicLetter);
        if (arabicError is not null)
        {
            return new PairResolution(null, arabicError);
        }
        return latin == arabic
            ? new PairResolution(null, AppMessages.LayoutsSame)
            : new PairResolution(new LayoutPair(latin, arabic), null);
    }

    private static (uint Hkl, string? Error) Pick(
        string setting, IReadOnlyList<(uint Hkl, string KeyA)> installed, Func<char, bool> isScriptLetter)
    {
        if (setting == "auto")
        {
            var matches = installed.Where(layout => layout.KeyA.Length == 1 && isScriptLetter(layout.KeyA[0])).ToList();
            return matches.Count == 1 ? (matches[0].Hkl, null) : (0, AppMessages.LayoutsNotFound);
        }
        var hkl = Hkl.Parse(setting);
        return installed.Any(layout => layout.Hkl == hkl) ? (hkl, null) : (0, AppMessages.LayoutNotInstalled(setting));
    }

    /// <summary>The same test as TextConverter: U+0600–U+06FF with a letter category.</summary>
    private static bool IsArabicLetter(char c) =>
        c is >= '؀' and <= 'ۿ'
        && CharUnicodeInfo.GetUnicodeCategory(c) is UnicodeCategory.UppercaseLetter
            or UnicodeCategory.LowercaseLetter
            or UnicodeCategory.TitlecaseLetter
            or UnicodeCategory.ModifierLetter
            or UnicodeCategory.OtherLetter;
}
